repo-updater: Convert Sources to send results on channel instead of returning a slice of repos

mrnugget opened the pull request:
This is the preparation for solving https://github.com/sourcegraph/sourcegraph/issues/5145.

It converts the interface of all existing `Source`s to send results on a channel. It keeps the existing behavior in the separate Sources regarding error handling: some accumulate errors and some return early.

The next step is then to use the channel in the syncer.

For ease of reviewing I think it makes sense to merge this here first.

keegancsmith commented on cmd/repo-updater/repos/syncer.go:
```diff
@@ -329,10 +329,33 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
+	done := make(chan struct{})
+	go func() {
+		srcs.ListRepos(ctx, results)
+		done <- struct{}{}
```
minor: A more common way I have seen this done is to `close(done)`. Will also catch double free then with a panic instead of a deadlock.

keegancsmith commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -81,11 +81,17 @@ const sourceTimeout = 30 * time.Minute
 
+type SourceResult struct {
+	Source Source
+	Repo   *Repo
```
We may implicitly code the Source in Repo as it is, but not sure what the implications of that are since we can have more than one source.

tsenart replied:
> We may implicitly code the Source in Repo as it is, but not sure what the implications of that are since we can have more than one source.

I think a good constraint to have is that a `SourceResult` always has a Repo with a single entry in the `Sources` field, corresponding to the `Source` field in the struct.

tsenart commented on cmd/repo-updater/repos/observability.go:
```diff
@@ -98,14 +98,31 @@ func NewSourceMetrics() SourceMetrics {
+	for res := range uncounted {
+		results <- res
+		if res.Err != nil {
+			err = res.Err
```
What do you think of aggregating all errors in a multierror?

tsenart commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -81,11 +81,17 @@ const sourceTimeout = 30 * time.Minute
 	ExternalServices() ExternalServices
 }
 
+type SourceResult struct {
```
Please add a Go doc.

tsenart commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -235,3 +210,30 @@ func group(srcs []Source) map[string]Sources {
 	return groups
 }
+
+func Drain(ctx context.Context, src Source) ([]*Repo, error) {
```
Please add a Go doc.

tsenart replied:
Would this read better as `ListAll`? Examples: `rs, err := repos.ListAll(ctx, src)`

mrnugget replied:
Yes, it would :) Thanks!

tsenart commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -235,3 +210,30 @@ func group(srcs []Source) map[string]Sources {
+		close(done)
+	}()
+
+	go func() {
```
Do need this extra go-routine or can we `close(results)` in the same go-routine as we `close(done)` (i.e. after we `close(done)`)?

mrnugget replied:
Ahh, good catch. With my original debugging/testing code gone, I guess this can be made simpler.

tsenart commented on cmd/repo-updater/repos/syncer.go:
```diff
@@ -325,7 +325,32 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
 	defer cancel()
 
-	return srcs.ListRepos(ctx)
+	results := make(chan *SourceResult)
```
What prevents us from using `Drain` here?

mrnugget replied:
Nothing really. I only chose to not use `Drain` because it currently does not build up `SourceError`s. Will try to refactor that out.

mrnugget commented:
@keegancsmith @tsenart I've addressed your comments/suggestions. Can you take a look again? 

Also, what's your opinion on merging this before the syncer has been switched to streaming?

keegancsmith commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -80,12 +80,24 @@ const sourceTimeout = 30 * time.Minute
-	ListRepos(context.Context) ([]*Repo, error)
+	// ListRepos sends all the repos a source yields over the passed in channel
+	// as SourceResults
+	ListRepos(context.Context, chan *SourceResult)
```
I am unsure of the implication, but maybe a `chan SourceResult` would result in less work for the GC? SourceResult type is reasonably small (two interfaces and a pointer).

mrnugget replied:
Yeah, I thought the same thing, then forgot about it. Thanks to your reminder I now went ahead and got rid of the `*` :)

keegancsmith commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -162,42 +142,17 @@ func (srcs Sources) ListRepos(ctx context.Context) ([]*Repo, error) {
-					ch <- result{src: src, errs: errs}
-				} else {
-					ch <- result{src: src, repos: repos}
-				}
```
woah nice simplification

keegancsmith commented on cmd/repo-updater/repoupdater/server.go:
```diff
@@ -311,7 +312,7 @@ func (s *Server) handleExternalServiceSync(w http.ResponseWriter, r *http.Reques
 		}
 
-		_, err = src.ListRepos(ctx)
+		_, err = repos.ListAll(ctx, src)
```
instead of using this helper, could we not just read over the channel and return the first error we see? that way we get back to the admin sooner. I suppose it is a single source, and all our sources stop after the first error so maybe this isn't necessary?

Edit: Do all our sources stop straight away? Maybe things like GitHub with repos, orgs, repoqueries won't. Will check.

mrnugget replied:
That's a good point, but not all sources stop after the first error. I quote my original PR message:

> It keeps the existing behavior in the separate Sources regarding error handling: some accumulate errors and some return early.

mrnugget replied:
Changed to return errors early, see latest commit

keegancsmith commented on cmd/repo-updater/repoupdater/server.go:
```diff
@@ -451,20 +452,24 @@ func (s *Server) handleStatusMessages(w http.ResponseWriter, r *http.Request) {
-					},
-				})
+		if multiErr, ok := errors.Cause(e).(*multierror.Error); ok {
+			for _, e := range multiErr.Errors {
```
unrelated: It's a shame multiErr doesn't satisfy `interface{Errors() []error}`.

keegancsmith commented on cmd/repo-updater/repos/testing.go:
```diff
@@ -17,14 +18,14 @@ import (
 func NewFakeSourcer(err error, srcs ...Source) Sourcer {
 	return func(svcs ...*ExternalService) (Sources, error) {
-		errs := new(MultiSourceError)
+		errs := new(multierror.Error)
```
minor: you don't have to allocate. you could just do `var errs multierror.Error`, multierror handles nil.

keegancsmith commented on cmd/repo-updater/repos/syncer.go:
```diff
@@ -325,12 +325,13 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
 func (s *Syncer) setOrResetLastSyncErr(perr *error) {
 	var err error
 	if perr != nil {
+		fmt.Printf("setOrResetLastSyncErr. *perr=%T, errors.Cause(*perr)=%T\n", *perr, errors.Cause(*perr))
```
debug

keegancsmith commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -235,3 +190,31 @@ func group(srcs []Source) map[string]Sources {
+
+// ListAll calls ListRepos on the given Source and collects the SourceResults
+// the Source sends over a channel into a slice of *Repo and a single error
+func ListAll(ctx context.Context, src Source) ([]*Repo, error) {
```
this doesn't need to be exported anymore, no other package uses it.

keegancsmith commented on cmd/repo-updater/repoupdater/server.go:
```diff
@@ -311,11 +312,26 @@ func (s *Server) handleExternalServiceSync(w http.ResponseWriter, r *http.Reques
+		for res := range results {
+			if res.Err != nil {
+				err = res.Err
+				break
```
you need to drain results.

mrnugget replied:
Ahh, good eye! What did you have in mind? Launching another goroutine that drains the channel?

mrnugget commented:
Just synced with @tsenart on this PR.

I'm going to go ahead and
1. merge this PR
2. follow-up with a PR that makes the sources `ctx.Done` aware, so that we don't have to worry about draining the `results` channel but can just cancel the `ctx` instead
3. follow-up with a PR that makes every source behave the same regarding "return early in case of error" or "accumulate errors and only return once completely done". Accumulating and continuing doesn't make a lot of sense in a single source (most likely the error will be the same for every request), we'd go with the "return early in case of error" version

keegancsmith commented:
> follow-up with a PR that makes the sources ctx.Done aware, so that we don't have to worry about draining the results channel but can just cancel the ctx instead

you still need to drain. Also with this approach everywhere you send down the channel you need to do a select with ctx.Done. That seems verbose and error prone.

mrnugget commented:
> you still need to drain. Also with this approach everywhere you send down the channel you need to do a select with ctx.Done. That seems verbose and error prone.

Agree on the "verbose on error prone" — I'm not too set on the `select` solution.

Do think it's okay as it is? Note that I added a "draining" of results.

keegancsmith commented:
LGTM
